### PR TITLE
[RoE] Implement weekly objectives, sparks exchange limit, voucher storage and exchange

### PR DESCRIPTION
**Pull request by ghost**
**_I affirm:_**
- [x] that I agree to Topaz Next's [Limited Contributor License Agreement](https://github.com/topaz-next/topaz/blob/release/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have read the [Contributing Guide](https://github.com/topaz-next/topaz/blob/release/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/topaz-next/topaz/blob/release/CODE_OF_CONDUCT.md)
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

**_Temporary_**:
- [x] that I understand there are large refactoring efforts going on right now, that these efforts touch every single Lua script and binding, and that my pull request might get put on hold to ensure there are not any conflicts with the refactoring work

This implements a lot of various items, which are as follows:

**Records of Eminence NPCs**
All Records of Eminence NPCs will now accept and store copper vouchers.  In addition, copper vouchers can now be exchanged for both currencies and provisions.

_Note: At the current time, Escha Silt is not implemented in the database for storage.  Current NPC behavior is to remove the currency, though no points will be awarded.  The map server will report a database error, should a player attempt to purchase it!_

**Weekly Sparks Exchange Limit**
Exchange limit has been implemented, and is flag configurable.  Current settings follow default retail behavior (enabled, maximum limit of 100000 sparks exchanged).

_From settings.lua:_
```
ENABLE_EXCHANGE_LIMIT = 1 -- Enable Maximum limit of sparks spent per Week (default retail behavior: 1)
WEEKLY_EXCHANGE_LIMIT = 100000 -- Maximum amount of sparks that can be spent per week (default retail value: 100000)
```

**Currency Caps**
Currency caps have been added for Unity Accolades, Ballista Points, Sparks of Eminence, and Valor Points.  Currently, this is only implemented for the sparks vendor, but `CLuaBaseEntity::addCurrency()` now supports an optional argument to allow passing a maximum limit.

_From settings.lua_
```
-- Currency Caps (Change at your own risk!)
CAP_CURRENCY_ACCOLADES = 99999
CAP_CURRENCY_BALLISTA  = 2000
CAP_CURRENCY_SPARKS    = 99999
CAP_CURRENCY_VALOR     = 50000
```

**Weekly RoE Objectives**
Weekly RoE objectives have been implemented, and similar to the exchange limit, are reset at JP midnight on the week transition.  Weekly objectives can be declared by setting the flag `"weekly"` within the RoE record.  A sample has been provided for RoE ID #3367 - "Signet, brb (W)"

_Note: Many weekly objectives have triggers that are not common to other objectives.  For this sample, `onRecordTrigger` is called directly in conquest.lua.  This may change over time as common groupings are discovered._

**Core Changes**
Outside of roe-specific changes to support resets, the following common functions have been modified, but should have no impact on previous function:

`CLuaBaseEntity::addCurrency(std::string const& currencyType, int32 amount, sol::object const& maxObj)`

addCurrency now supports an optional cap to be passed to `charutils::AddPoints()`.  No argument, or a nil value passed will function as previously implemented.

`CLuaBaseEntity::setCharVar()` database function has been moved to `charutils::SetCharVar()` which the binding now references.

### CODE REVIEW

File: src/map/time_server.cpp [hunk 1]
```diff
@@ -51,6 +51,8 @@ int32 time_server(time_point tick, CTaskMgr::CTask* PTask)
             conquest::UpdateWeekConquest();
             lastConquestTally = tick;
         }
+
+        roeutils::CycleWeeklyRecords();
```
Comment (cocosolos) [inline src/map/time_server.cpp]: Should this be moved into the block above? I think this will get hit several times.

Comment (ghost) [reply src/map/time_server.cpp]: Thanks for the feedback!  The 1h offset threw me originally, but now I understand its purpose.

File: scripts/zones/Bastok_Markets/IDs.lua [hunk 2]
```diff
@@ -66,7 +67,10 @@ zones[tpz.zone.BASTOK_MARKETS] =
         THANK_YOU                    = 8279, -- Thank you...
         RETRIEVE_DIALOG_ID           = 12885, -- You retrieve <item> from the porter moogle's care.
         TURNING_IN_SPARKS            = 14206, -- Ohohoho... Turning in sparks, I see.
+        DO_NOT_POSSESS_ENOUGH        = 14230, -- You do not possess enough <item name> to complete the transaction.
```
Comment (zach2good) [inline scripts/zones/Bastok_Markets/IDs.lua]: I seem to remember that the syntax of the comments here are somewhat important. Check in with @wrenffxi to see if this needs to be `<item>`

File: src/map/lua/lua_baseentity.cpp [hunk 3]
```diff
@@ -464,19 +464,11 @@ int32 CLuaBaseEntity::getCharVar(std::string const& varName)
...
-    {
-        Sql_Query(SqlHandle, "DELETE FROM char_vars WHERE charid = %u AND varname = '%s' LIMIT 1;", m_PBaseEntity->id, varname.c_str());
-    }
-    else
-    {
-        const char* fmtQuery = "INSERT INTO char_vars SET charid = %u, varname = '%s', value = %i ON DUPLICATE KEY UPDATE value = %i;";
-        Sql_Query(SqlHandle, fmtQuery, m_PBaseEntity->id, varname.c_str(), value, value);
-    }
+    auto* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity);
+    return charutils::SetCharVar(PChar, varName.c_str(), value);
```
Comment (zach2good) [inline src/map/lua/lua_baseentity.cpp:471]: 👍

File: src/map/lua/luautils.cpp [hunk 4]
```diff
@@ -4128,6 +4128,8 @@ namespace luautils
     {
         TracyZoneScoped;
 
+        ShowDebug("LoadEventScript: Checking Function %s\n", functionName);
```
Comment (zach2good) [inline src/map/lua/luautils.cpp]: I'm in favour of these messages, but perhaps not for Debug. I've been considering if we need `ShowDev` for REALLY extended logging...

**Review by zach2good — APPROVED** [review]

File: src/map/lua/luautils.cpp (hunk 4 above)
Comment (ghost) [reply src/map/lua/luautils.cpp]: Accidentally check it back in after it was removed.  Removed again to match the current base.

File: scripts/zones/Bastok_Markets/IDs.lua (hunk 2 above)
Comment (ghost) [reply scripts/zones/Bastok_Markets/IDs.lua]: Updated comments based on Wren's feedback.  I _hope_ I didn't break anything!

**Review by zach2good — APPROVED** [review]

**Pull request merged**